casper/src/main/scala/io/casperlabs/casper/util/execengine/ExecEngineUtil.scala

```diff
@@ -99,90 +82,258 @@ object ExecEngineUtil {
+    def empty: TransformMap                                       = Nil
+  }
+
+  sealed trait MergeResult[+T, +A] { self =>
+    def parents: Vector[A] = self match {
+      case MergeResult.EmptyMerge            => Vector.empty[A]
+      case MergeResult.Result(head, _, tail) => head +: tail
+    }
+
+    def transform: Option[T] = self match {
+      case MergeResult.EmptyMerge      => None
+      case MergeResult.Result(_, t, _) => Some(t)
+    }
+  }
+  object MergeResult {
+    case object EmptyMerge extends MergeResult[Nothing, Nothing]
+    case class Result[T, A](
+        firstParent: A,
+        nonFirstParentsCombinedEffect: T,
+        nonFirstParents: Vector[A]
+    ) extends MergeResult[T, A]
+
+    def empty[T, A]: MergeResult[T, A] = EmptyMerge
+    def result[T, A](
+        firstParent: A,
+        nonFirstParentsCombinedEffect: T,
+        nonFirstParents: Vector[A]
+    ): MergeResult[T, A] = Result(firstParent, nonFirstParentsCombinedEffect, nonFirstParents)
+  }
+
+  /** Computes the largest commuting sub-set of blocks from the `candidateParents` along with an effect which
+    * can be used to find the combined post-state of those commuting blocks.
+    * @tparam F effect type (a la tagless final)
+    * @tparam T type for transforms (i.e. effects deploys create when executed)
+    * @tparam A type for "blocks". Order must be a topological order of the DAG blocks form
+    * @tparam K type for keys specifying what a transform is applied to (equal to ipc.Key in production)
+    * @param candidates "blocks" to attempt to merge
+    * @param parents function for computing the parents of a "block" (equal to _.parents in production)
+    * @param effect function for computing the transforms of a block (looks up the transaforms from the blockstore in production)
+    * @param toOps function for converting transforms into the OpMap, which is then used for commutativity checking
+    * @return an instance of the MergeResult class. It is either an `EmptyMerge` (which only happens when `candidates` is empty)
+    *         or a `Result` which contains the "first parent" (the who's post-state will be used to apply the effects to obtain
+    *         the merged post-state), the combined effect of all parents apart from the first (i.e. this effect will give
+    *         the combined post state for all chosen commuting blocks when applied to the post-state of the first chosen block),
+    *         and the list of the additional chosen parents apart from the first.
+    */
+  def abstractMerge[F[_]: Monad, T: Monoid, A: Ordering, K](
+      candidates: IndexedSeq[A],
+      parents: A => F[List[A]],
+      effect: A => F[Option[T]],
+      toOps: T => OpMap[K]
+  ): F[MergeResult[T, A]] = {
+    val n = candidates.length
+
+    def netEffect(blocks: Vector[A]): F[T] =
+      blocks
+        .traverse(block => effect(block))
+        .map(_.flatten.foldLeft[T](Monoid[T].empty)(Monoid[T].combine))
```

Comment:
```suggestion
        .map(_.flatten.fold)
```

`fold` is defined on `cats.Foldable` and it's basically what you have here: `foldLeft + Monoid`.

Reply:
Nice!
Done